### DISCUSSION TITLE
General app behavior when foregrounding app from Notifications

### DISCUSSION THREAD

**Discussion by munkii** [shinyorg/shiny#1478] [Q&A]
This is not about an issue with Shiny but it is related to things I am doing with my app that Shiny helps me with and I thought I'd ask the Shiny community a question.

@aritchie if this is not what you want in the Q&A please just close it.

Do you simply restart your app when a Notification comes in?

The standard approach that I have seen for XF apps is the following. In your Xamarin.Forms.Application OnStart, initialise your Navigation Page/Manager Stack and load your Main/Home View or ViewModel. We are currently using FreshMVVM although trying to migrate to Prism for MAUI. 

Ours looks something like this.

```
protected override void OnStart()
{
    base.OnStart();

    Page pageToShow = null;

    if (Xamarin.Essentials.VersionTracking.IsFirstLaunchEver == false)
    {
        pageToShow = FreshPageModelResolver.ResolvePageModel<HomeViewModel>();
    }
    else
    {
        pageToShow = FreshPageModelResolver.ResolvePageModel<FirstRunViewModel>();
    }

    var mainNavContainer = new FreshNavigationContainer(pageToShow, NavigationContainerNames.MainContainer);
    this.MainPage = mainNavContainer;
}
```

What this standard approach seems to mean in relation to Notifications is that when your app is resumed from a Notifcation tap this code fires and users are taken back to Main/Home every time no matter where they were in the app.

Is that what other people are doing?

I know that we could do some extra work in Shiny's PushDelgate.OnEntry to move the user onwards to a notification specfic View but the Navigation stack would be still have been reset to Home/Main first.

**Comment by aritchie** [comment]
I don't do anything in App.OnStart.  I push everything to a startup page where I have a full nav stack and the ability to control history properly.  OnEntry should only be pushing an event or setting some variable for startup to deal with.

**Reply by munkii** [reply]
Do you do anything different between OnReceived and OnEntry?

I have always been trying to to interpret OnEntry as "User has tapped on notification, take them where they need to go". That was dumb. I will look at switching to just logging "an event or setting some variable for startup to deal with"

BTW SQIite Logging. Love it, great for debugging Delegates thanks

**Reply by aritchie** [reply]
Yes

OnReceived is used general for "silent" stuff like triggering an HTTP call or running a job... stuff like that
OnEntry is exactly what you stated... "User has tapped, go somewhere"

Glad SQLite is working out for you. It became a necessity for local debugging

**Reply by munkii** [reply]
I ask because under some circumstances such as 

- Android
- App not Running
- User taps on Notification

I am **not** getting an OnEntry only an OnReceived so loading OnEntry with the responsibility of actioning a Navigation was an issue as in this scenario it never happened. App just loaded and PN action was ignored.

I'm planning to simply stash any parsed data during OnReceived, in OnEntry I will double check that the Notifcation data has been stashed and then in our StartUp page I will action the PN

**Reply by aritchie** [reply]
I need something reproducible in order to suggest why it may not be working

**Reply by munkii** [reply]
I have it working as I'd expect in our MAUI branch so I'm happy with that for now.

I'm going to switch from dealing/processing notifcations in my Delegates to simply packaging them up and then Publishing an event so they can be dealt with elsewhere.

Taking notification to "navigate the user to a page" as an example, where do you listen for those kind of events? I was just going to put the event subscription in our Homepage as we always have one and only one. I know it will "work" but seems kind of wrong, like i'm overloading that page with responsibility (again)
